### ISSUE TITLE
Finagle load balancers are hard to use

### ISSUE THREAD

**Issue by scf37** [twitter/finagle#424]
I do not want to use balancer blindly as it is very important part of service stability, so Httpx.newService("host1, host2") is not an option.

But - to add balancer for manually managed pool of Service instances is insane amount of work. You need to wrap them to ServiceFactory, remember about Service pooling issues, convert balancer instance  back to Service.

Altering StackClient's stack? Stack has a lot of components with their params scattered around the code - there is no easy answer to "what parameters can I use to configure this Stack?". Also, understanding of how those components are configured (more complex scattered code) and inter-operate (they mostly independent, but not totally) is hard.

Hell, I better compose filters and factories I need via andThen but... that is not easy task as well.

To me, idea of Service <-> ServiceFactory translations throughout req/resp processing pipeline seems to be a mistake.

**Comment by mosesn** [comment]
@scf37 One of the advantages of open source software is that you're using software (like load balancers) that have been tested extensively by other companies, including at scale.  Could you elaborate on why you don't want to use any of the load balancers that we already provide?

The ServiceFactory / Service abstraction generally works well for us, although there are a few places where we reconvert from Service to ServiceFactory, which I agree can be confusing.  We're working on smoothing it out.

**Comment by mariusae** [comment]
Could you elaborate on why you can't use the default load balancing configuration?

> To me, idea of Service <-> ServiceFactory translations throughout req/resp processing pipeline seems to be a mistake.

Can you elaborate? This is very much by design--load balancing occurs over *sessions* and not requests, as many protocols require this.

**Comment by scf37** [comment]
@mariusae As for default configuration: Problem is default configuration is large and loosely documented. I do not need just 'balancing', I need to build system that can withstand different kinds of failures within time limits. and therefore I need to understand how balancing/retrying/error classification and handling works. As far as my experience tells me, there is no 'best' configuration - most is a trade-off between load , failure impact and recovery speed. Understanding defaults fully is tough and they could be changed in the future. So I prefer to combine filters manually.

As for design. I may be wrong here as I do not know Finagle internals well enough, so...
Rants:

- Given Var[Set[Service]], create Service that balances that set. This task is too complex - you'll need to wrap/unwrap service factories and configure pools to keep underlying services opened.
- Entire Stack implementation looks like typed dependency injection framework. While it allows to combine, re-combine and configure any part, it requires additional code to insert even simple filter w/o parameters. Also, combined Stack is hard to inspect and modify: What filters are in? What parameters do they have? What parameters are set? Can Stack type contain information on every element it contains?

Suggestions:
- Balancers need better isolation from what they balance. It looks like right now balancer client must call ServiceFactory.close() to indicate he is done with that connection/session/request so it can be returned to balancer's pool. Why not provide abstraction that can both work with Service and ServiceFactory and both with one-time and persistent connections? So programmer do not need to protect Balancer pool services from being really closed 'by design'.
- Stack needs better toString and ability to enumerate available properties for every attached module. May be more. I was thinking of possibility to make Stack-s by combining traits so we get something typed in the end... Or may be something else.
- ClientBuilder/ServerBuilder are great. Really.They list all available configuration in single interface so one can see what they do, what they don't do and what can be configured. So they are ideal for novices (give me something working but configurable) and experts looking for typical stack.
- What I love in Finagle most is that it is (mostly) composable. It "don't want it - just don't use it" ideology. So I'd love to see better support for building finagle pipelines from scratch or predefined parts (like dam balancers with dam pooling and dam wrappers)

Phew. 

I really like Finagle. Both its core ideology and dozens of technological marvels buried deep in the code, teaching valuable lessons on distributed services to anyone persistent enough to find them.

I still have a lot to learn so take above wall of text lightly, it is just an opinion :-) And, guys, write better documentation! Finagle is much more than "another simple HTTP client" or "low-level HTTP server" but who knows about that?

**Comment by mosesn** [comment]
> Given Var[Set[Service]], create Service that balances that set. This task is too complex - you'll need to wrap/unwrap service factories and configure pools to keep underlying services opened.

Why do you want to do this?  The point of having StackClient is so that you don't have to do this.  Connections pools keep persistent connections, so there's no reason you have to do this manually.

> Entire Stack implementation looks like typed dependency injection framework. While it allows to combine, re-combine and configure any part, it requires additional code to insert even simple filter w/o parameters. Also, combined Stack is hard to inspect and modify: What filters are in? What parameters do they have? What parameters are set? Can Stack type contain information on every element it contains?

If you have to modify the Stack, either you're doing something incorrectly, or we should probably support your use case in StackClient.  It's probably not the API you should be using, and we tried to put warning in the API docs.  See ["Note: Stacks are advanced and sometimes subtle. For expert use only!"][0].

With that said, Stack is pretty easy to inspect.  It has a nice [Stack#toString][1] implementation.

Yes, and in fact we do expose all of this data.  If you use twitter-server, you can hit /admin/registry.json and it explains every level of the stack and how each level has been configured.

>Balancers need better isolation from what they balance. It looks like right now balancer client must call ServiceFactory.close() to indicate he is done with that connection/session/request so it can be returned to balancer's pool. Why not provide abstraction that can both work with Service and ServiceFactory and both with one-time and persistent connections? So programmer do not need to protect Balancer pool services from being really closed 'by design'.

We do, they're connection pools. See WatermarkPool, SingletonPool, DefaultPool, etc.  When the LoadBalancer says, "close this session" it means, "I no longer need this connection, so somebody else can use it."  It doesn't mean it cuts the connection.  Finagle doesn't natively support any protocols where you cut the connection, so if you're seeing this behavior with StackClient, without modifying the Stack, I'd be very surprised.

> Stack needs better toString and ability to enumerate available properties for every attached module. May be more. I was thinking of possibility to make Stack-s by combining traits so we get something typed in the end... Or may be something else.

I think the toString method might actually do do what you want it to do.  Could you elaborate on what you expect it to do versus what you want it to do?

> ClientBuilder/ServerBuilder are great. Really.They list all available configuration in single interface so one can see what they do, what they don't do and what can be configured. So they are ideal for novices (give me something working but configurable) and experts looking for typical stack.

We're aware of the problem with the finagle 6 APIs, and working on fixing it.

> What I love in Finagle most is that it is (mostly) composable. It "don't want it - just don't use it" ideology. So I'd love to see better support for building finagle pipelines from scratch or predefined parts (like dam balancers with dam pooling and dam wrappers)

We've been thinking about this, but I'd be very curious about your use case.

@scf37 it feels like you're coming into the project first to criticize it, and it might be more productive if you come in first to try to understand it.  For example, a lot of these questions where you misunderstood what finagle was doing could have been framed as something like, "I don't understand why it seems like load balancers cut the underlying connections after they're done, but we also have connection pools–those seem contradictory.  What am I missing?"

My main gripe is that it feels like you're on the attack in our open source community, and I don't want to feel like I have to defend myself against folks who are really on the same side as me–it sounds like you also want to make finagle better, which is great!  But I'm afraid I'm probably not going to enjoy working with you to make finagle be the best it can be if all I see from you is, "Finagle is terrible because I don't understand it and think it does really dumb things everywhere."  This is especially frustrating because I feel like I already answered some of these questions for you on the listserv, but it seems like you didn't read my responses.

[0]: https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/Stack.scala#L19
[1]: https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/Stack.scala#L198-L204

**Comment by luciferous** [comment]
FWIW I didn't see @scf37's comments as an attack. I thought the amount of detail in them demonstrated understanding and that a lot of thought (and usage) went into them.

**Comment by scf37** [comment]
@mosesn Please do not see this as attack. I'm just providing input so both me and you can learn. It is because negative input is the most helpful - you obviously should be more interested in parts your users aren't happy with while I surely can benefit from maintainer's point of view on Finagle. Collaboration is working together at the same goal - but I'm not there yet, I do not know enough of use cases and maintainer's intentions. So, I try to voice what I see as a problem so, again, you may know what users see a problem while I get your responses for my concerns so I can learn.

To summarize, you stand on "this can be done" and "you do not need to use low level APIs like Stack". While I'm all about "this is too complex" and "I want to control behavior so I want to use low level APIs"

Or to say it in other way, I have no interest in framework that "just works". I'm interested in ideas, small independent components and their composition. You see, after 5 years with Hibernate/JSF I have no more faith in frameworks :-)

P.S. Looking again at Stack.toString - still only role name and description is printed.

**Comment by mosesn** [comment]
OK, thanks for explaining your rationale.  I think we do want to make it *possible* to construct your own Stacks, but it's unlikely that we're going to put a lot of effort into making it easy.  Our focus is on making the common path easy, and the build-from-scratch path possible.  But if there are simple ways of making the build-from-scratch path better without hurting the common path, of course we'd be happy to hear about them.

If you want information on parameters, right now you can get it from TwitterServer's /admin/registry.json.  We don't expose that in Stack#toString yet, but happy to accept a pull request.

**Comment by mariusae** [comment]
Hi @scf37, thanks for the details! Some replies inline:

> Given Var[Set[Service]], create Service that balances that set. This task is too complex - you'll need to wrap/unwrap service factories and configure pools to keep underlying services opened.
...
> Balancers need better isolation from what they balance. It looks like right now balancer client must call ServiceFactory.close() to indicate he is done with that connection/session/request so it can be returned to balancer's pool. Why not provide abstraction that can both work with Service and ServiceFactory and both with one-time and persistent connections? So programmer do not need to protect Balancer pool services from being really closed 'by design'.

What you're describing is almost exactly the current design, it's just that that generic API  is `ServiceFactory`! All a `ServiceFactory` is is a way to produce a service, and to give it back. For example, `ServiceFactory.const` gives you a `ServiceFactory` that just uses a single service. In your case, 

    val services: Var[Set[Service[..]]] = ...
    val factories: Var[Set[ServiceFactory[..]]] = services.map(svcs => svcs.map(ServiceFactory.const))

That's really all there's to it.

> Entire Stack implementation looks like typed dependency injection framework. While it allows to combine, re-combine and configure any part, it requires additional code to insert even simple filter w/o parameters. Also, combined Stack is hard to inspect and modify: What filters are in? What parameters do they have? What parameters are set? Can Stack type contain information on every element it contains?

You're totally right. This is still an "expert level" API. The idea behind Stack is to treat Finagle's Stack as any other collection, which you can transform, inspect, instantiate, etc... I think the idea is very solid, but the implementation is still a little cumbersome and difficult to use. This is something the team at Twitter is actively working on improving.

**Comment by mosesn** [comment]
I'm closing this ticket, since it's more feedback than a well-formed issue.  The issues that @scf37 raised are something we've been thinking about, and down the line, we would like to improve the expert-level API.
